### PR TITLE
Project 20 - Shaun Newmark and Erika Andersson Porath

### PR DESCRIPTION
**Pull request by newmarkdev**

### CODE REVIEW

File: frontend/src/components/Header.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import { batch, useSelector, useDispatch } from 'react-redux';
+import Button from '@material-ui/core/Button';
+
```
Comment (Marielle8) [inline frontend/src/components/Header.js:4]: Remove an empty line

File: frontend/src/components/Login.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    const [password, setPassword] = useState('');
+    const [mode, setMode] = useState(null);
+
+    const accessToken = useSelector(store => store.user.accessToken);
+    const dispatch = useDispatch();
+    const history = useHistory();
+
+    useEffect(() => {
+        // redirect user to '/' path
+        console.log('Checking access token', accessToken);
```
Comment (Marielle8) [inline frontend/src/components/Login.js:51]: Remove console.log

File: frontend/src/components/Login.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            method: 'POST',
+            headers: {
+                'Content-Type': 'application/json'
+            },
+            body: JSON.stringify({ username, password })
+        }
+        fetch(API_URL(mode), options)
+            .then(res => res.json())
+            .then(data => {
+                console.log(data);
```
Comment (Marielle8) [inline frontend/src/components/Login.js:70]: Remove console.log

File: frontend/src/reusable/urls.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+const BASE_URL = 'http://localhost:8080';
```
Comment (Marielle8) [inline frontend/src/reusable/urls.js:1]: This url should be the one from your deployed backend in Heroku.

**Review by Marielle8 — CHANGES_REQUESTED** [review]
This is overall a good project with very nice styling and clean code that is easy to follow! Great work with that 😀🙏! But we can't see that you  return a 401 or 403 with an error message if you try to access without an authentication which was one of the blue requirments for this week?

File: README.md [hunk 5]
```diff
@@ -1,13 +1,19 @@
...
+After the log in the user is redirected to an authorized site, where they also can log out and return on the log in screen. We used a Material UI library
+for styling.
+Our main problem was time management because it was difficult to find a time we could work together since we have different working hours. We ended up doing a lot on our own end since we both followed along and used the solutions in the lectures, and then we met up and merged our projects together. The errors we ran into was mostly caused by typos. If we had more time we would like to have improved the frontend with better shown error messages and let the user post things on the authorized side. On the backend signed we would like to improve the validations for the sign-up/sign-in form.
 
 ## View it live
 
-Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
+The project is deployed in Heroku and Netlify, and the database is set up in Mongo Cloud Atlas.
+See the deployed frontend here:  https://fervent-allen-86afba.netlify.app/
```
Comment (KarolCislowski) [inline README.md:19]: To fix it you need to create `_redirects` file in the `public` directory of your react project and paste it inside:
``` 
/* /index.html 200
```

image

File: backend/server.js [hunk 6]
```diff
@@ -1,30 +1,85 @@
...
+  },
+  password: {
+    type: String,
+    required: true
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString('hex')
+  }
+});
```
Comment (KarolCislowski) [inline backend/server.js:25]: Nice model 👍🏻

File: frontend/src/components/Header.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+const Header = () => {
+    const accessToken = useSelector(store => store.user.accessToken);
+    const dispatch = useDispatch();
+
+    const onButtonClick = () => {
+        batch(() => {
+            dispatch(user.actions.setUsername(null));
+            dispatch(user.actions.setAccessToken(null));
+        });
```
Comment (KarolCislowski) [inline frontend/src/components/Header.js:16]: 👍🏻

File: frontend/src/components/Header.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+    return (
+        <header>
+            {accessToken && <Button 
+            fullWidth
+            color="primary" 
+            onClick={onButtonClick}>Logout
+            </Button>}
+        </header>
+    );
```
Comment (KarolCislowski) [inline frontend/src/components/Header.js:27]: indentation should be set to 2 spaces.

File: frontend/src/reusable/urls.js (hunk 4 above)
Comment (KarolCislowski) [reply frontend/src/reusable/urls.js:1]: This is why nothing is working 😭

File: frontend/src/reducers/user.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+        setAccessToken: (store, action) => {
+            store.accessToken = action.payload;
+        },
+        setErrors: (store, action) => {
+            store.errors = action.payload;
+        }
+    }
+});
+
+export default user;
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:23]: You need to store your data in the local storage to avoid logging out your user after refreshing the page.

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
It is really nice app but you need to change the API URL in your react app. clean it up and persist your user in the local storage to avoid refreshing.


Because your graduation is so soon please let me know on slack (a short note with a link to this PR) when you finish fixing your app.

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
You still need to save your user data in the local storage to not lose it on refresh.

File: frontend/src/reducers/user.js (hunk 9 above)
Comment (newmarkdev) [reply frontend/src/reducers/user.js]: This should be added now 👍

**Review by KarolCislowski — APPROVED** [review]
Great Job!

**Pull request closed without merging**